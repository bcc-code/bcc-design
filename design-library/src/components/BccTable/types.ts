export type BccTableColumn = {
  key: string;
  text?: string;
  sortable?: boolean;
  sortMethod?: (a: BccTableItem, b: BccTableItem) => number;
};
export type BccTableItem = Record<string, any>;
export type BccTableSortDirection = "ascending" | "descending";
