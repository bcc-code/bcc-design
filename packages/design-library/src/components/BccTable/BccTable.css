@layer components {
    .bcc-table {
        @apply min-w-full;
    }

    /* Wrapper */
    .bcc-table-wrapper {
        @apply overflow-x-auto;
    }

    /* Head */
    .bcc-table-header-row {
        @apply bg-gray-50;
    }

    /* Header */
    .bcc-table-header {
        @apply text-label-small px-4 py-3 text-left whitespace-nowrap text-secondary;
        @apply first:rounded-l-lg last:rounded-r-lg;
    }
    .bcc-table-header-sortable {
        @apply cursor-pointer;
    }

    .bcc-table-header-content {
        @apply flex items-center gap-2;
    }

    .bcc-table-header-sort-icon {
        @apply h-5 w-5 text-tertiary;
    }
    .bcc-table-header-sort-icon-undetermined {
        @apply text-transparent;
    }
    .bcc-table-header-sortable:hover .bcc-table-header-sort-icon {
        @apply inline-block text-secondary;
    }

    /* Cell */
    .bcc-table-cell {
        @apply text-body-small px-4 py-3 whitespace-nowrap text-secondary;
    }
}
