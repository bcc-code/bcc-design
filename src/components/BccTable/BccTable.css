@layer components {
    .bcc-table {
        @apply w-full;
    }

    /* Head */
    .bcc-table-header-row {
        @apply bg-gray-50;
    }

    /* Header */
    .bcc-table-header {
        @apply text-label-small px-4 py-3 text-left text-secondary;
        @apply first:rounded-l-lg last:rounded-r-lg;
    }

    .bcc-table-header-content {
        @apply flex items-center;
    }

    .bcc-table-header-sort-icon {
        @apply h-5 w-5 text-tertiary;
    }

    /* Cell */
    .bcc-table-cell {
        @apply text-body-small px-4 py-3 text-secondary;
    }
}
