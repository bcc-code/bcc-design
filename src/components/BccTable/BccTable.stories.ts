import BccTable from "./BccTable.vue";
import BccButton from "@/components/BccButton/BccButton.vue";
import BccBadge from "@/components/BccBadge/BccBadge.vue";
import { ChevronRightIcon } from "@bcc-code/icons-vue";

import type { Meta, StoryFn } from "@storybook/vue3";

/**
 * A versatile table with built-in sorting.
 */
export default {
  title: "Components/BccTable",
  component: BccTable,
  argTypes: {
    sortDirection: {
      options: ["ascending", "descending"],
      control: { type: "radio" },
    },
  },
} as Meta<typeof BccTable>;

const Template: StoryFn<typeof BccTable> = (args) => ({
  components: { BccTable, BccButton, BccBadge },
  setup() {
    return { args, ChevronRightIcon };
  },
  template: `
    <BccTable :columns="args.columns" :items="args.items" v-model:sortBy="args.sortBy" v-model:sortDirection="args.sortDirection">
      <template #item.status="{ item }">
        <BccBadge :context="item.status.context">{{ item.status.text }}</BccBadge>
      </template>
      <template #item.actions="{ item }">
        <BccButton variant="tertiary" size="sm" :padding="false" :icon="ChevronRightIcon" iconRight>Evaluation</BccButton>
      </template>
    </BccTable>
  `,
});

const statusSort = (a, b) => {
  if (a.status.text > b.status.text) return 1;
  if (a.status.text < b.status.text) return -1;
  return 0;
};

/**
 * Pass a `columns` array with `text` (displayed as the column header in the table) and `key` (used to reference values) elements. Pass an `items` array with objects that have an item for each of the `key`s in the columns (or pass a slot for columns that do not have associated data, see below).
 *
 * By default the contents whatever is in the item key will be rendered. You can render something from deeper levels by using a dot notation, such as `progress.amount`.
 * Pass a slot to the table named `item.<column key>` to control the rendering. In this example both for `status` and `actions` a custom table cell is rendered. This slot is a [scoped slot](https://vuejs.org/guide/components/slots.html#scoped-slots) and receives the current `item` as a slot prop.
 *
 * Columns are sortable by default, set `sortable: false` on a column to disable sorting. Set a `sortMethod` on a column to have a custom sort method, handy when that particular column isn't something that can be sorted alphabetically or numerically by default (note that our example on the status column here only works with ascending sort due to Storybook limitations).
 */
export const Example = Template.bind({});
Example.args = {
  sortBy: undefined,
  sortDirection: "descending",
  columns: [
    {
      text: "Year group",
      key: "year_group",
    },
    {
      text: "Name",
      key: "name",
    },
    {
      text: "Status",
      key: "status",
      sortMethod: statusSort,
    },
    {
      text: "Progress",
      key: "progress.amount",
    },
    {
      text: "Actions",
      key: "actions",
      sortable: false,
    },
  ],
  items: [
    {
      id: 1,
      year_group: "23/24",
      name: "Ola Nordmann",
      status: { text: "On Track", context: "success" },
      progress: {
        amount: 2,
      },
    },
    {
      id: 2,
      year_group: "22/23",
      name: "Johan Oscar",
      status: { text: "Finished", context: "info" },
      progress: {
        amount: 25,
      },
    },
    {
      id: 3,
      year_group: "22/23",
      name: "Ada Lovelace",
      status: { text: "Behind Schedule", context: "warning" },
      progress: {
        amount: 15,
      },
    },
    {
      id: 4,
      year_group: "23/24",
      name: "Firmus Piett",
      status: { text: "On Track", context: "success" },
      progress: {
        amount: 15,
      },
    },
  ],
};
Example.parameters = {
  docs: {
    source: {
      language: "html",
      code: `
<BccTable
  :columns="columns"
  :items="items"
  v-model:sortBy="sortBy"
  v-model:sortDirection="sortDirection"
>
  <template #item.status="{ item }">
    <BccBadge :context="item.status.context">{{ item.status.text }}</BccBadge>
  </template>
  <template #item.actions="{ item }">
    <BccButton variant="tertiary" size="sm" :padding="false" :icon="ChevronRightIcon" iconRight @click="editItem(item.id")>Evaluation</BccButton>
  </template>
</BccTable>
    `,
    },
  },
};
