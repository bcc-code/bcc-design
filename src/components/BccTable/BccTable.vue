<script setup lang="ts">
import { SwapVertIcon, ArrowUpwardIcon, ArrowDownwardIcon } from "@bcc-code/icons-vue";
import { onMounted, ref, toRefs } from "vue";

type Column = {
  key: string;
  text?: string;
  sortable?: boolean;
  sortMethod?: (a: Item, b: Item) => number;
};
type Item = Record<string, any>;
type SortDirection = "ascending" | "descending";

type Props = {
  columns: Column[];
  items: Item[];
  sortBy?: string;
  sortDirection?: SortDirection;
};

const props = withDefaults(defineProps<Props>(), {
  sortDirection: "descending",
});
const { columns, items, sortBy, sortDirection } = toRefs(props);

const emit = defineEmits(["update:sortBy", "update:sortDirection"]);

let sortedItems = ref([...items.value]);

function resetSortedItems() {
  sortedItems.value = [...items.value];
}

function sortItems(column: Column, direction: SortDirection) {
  if (column.sortMethod) {
    sortedItems.value.sort(column.sortMethod);
    return;
  }

  sortedItems.value.sort((a, b) => {
    const firstItem = getField(a, column.key);
    const secondItem = getField(b, column.key);
    if (firstItem < secondItem) return direction == "descending" ? 1 : -1;
    if (firstItem > secondItem) return direction == "descending" ? -1 : 1;
    return 0;
  });
}

function sort(column: Column) {
  if (column.sortable === false) {
    return;
  }

  // Stop sorting this column if we have cycled through both sorting directions
  if (sortBy?.value == column.key && sortDirection.value === "descending") {
    resetSortedItems();
    emit("update:sortDirection", "descending");
    emit("update:sortBy", undefined);
    return;
  }

  // If the column was already sorted, flip the sort direction
  if (sortBy?.value == column.key) {
    emit("update:sortDirection", "descending");
    sortItems(column, "descending");
    return;
  }

  // Reset the sort direction when another column is clicked
  emit("update:sortDirection", "ascending");
  emit("update:sortBy", column.key);
  sortItems(column, "ascending");
}

function getField(item: Item, columnKey: string) {
  const segments = columnKey.split(".");
  return segments.reduce((obj, key) => obj[key], item);
}

onMounted(() => {
  if (sortBy?.value) {
    const column = columns.value.find((c) => c.key == sortBy.value);
    sortItems(column!, sortDirection.value);
  }
});
</script>

<template>
  <div class="bcc-table-wrapper">
    <table class="bcc-table">
      <thead>
        <tr class="bcc-table-header-row">
          <th
            class="bcc-table-header"
            :class="{ 'bcc-table-header-sortable': column.sortable !== false }"
            v-for="column in columns"
            :key="column.key"
            @click="sort(column)"
          >
            <div class="bcc-table-header-content">
              {{ column.text }}

              <button
                v-if="column.sortable !== false"
                aria-label="Sort column"
                class="bcc-table-header-sort"
              >
                <SwapVertIcon
                  class="bcc-table-header-sort-icon bcc-table-header-sort-icon-undetermined"
                  v-if="sortBy !== column.key"
                />
                <ArrowUpwardIcon
                  class="bcc-table-header-sort-icon"
                  v-if="sortBy == column.key && sortDirection == 'ascending'"
                />
                <ArrowDownwardIcon
                  class="bcc-table-header-sort-icon"
                  v-if="sortBy == column.key && sortDirection == 'descending'"
                />
              </button>
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr v-for="(item, key) in sortedItems" :key="key">
          <td class="bcc-table-cell" v-for="column in columns" :key="column.key">
            <slot v-if="$slots[`item.${column.key}`]" :name="`item.${column.key}`" :item="item" />
            <span v-else>
              {{ getField(item, column.key) }}
            </span>
          </td>
          <td></td>
        </tr>
      </tbody>
    </table>
  </div>
</template>
