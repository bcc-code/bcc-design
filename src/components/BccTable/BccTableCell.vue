<template>
  <td class="bcc-table-cell">
    <slot />
  </td>
</template>
