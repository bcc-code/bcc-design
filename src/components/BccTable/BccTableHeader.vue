<template>
  <th class="bcc-table-header">
    <slot />
  </th>
</template>
