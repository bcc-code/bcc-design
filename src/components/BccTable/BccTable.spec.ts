import { describe, it, expect } from "vitest";

import { mount } from "@vue/test-utils";
import BccTable from "./BccTable.vue";

const columns = [
  {
    text: "Year group",
    key: "year_group",
  },
  {
    text: "Name",
    key: "name",
  },
];
const items = [
  {
    year_group: "22/23",
    name: "Ada Lovelace",
  },
  {
    year_group: "23/24",
    name: "Firmus Piett",
  },
];

describe("BccTable", () => {
  it("renders data in a table", () => {
    const wrapper = mount(BccTable, {
      props: {
        columns,
        items,
      },
    });

    expect(wrapper.text()).toContain("Year group");
    expect(wrapper.text()).toContain("Ada Lovelace");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("can render a custom table cell", () => {
    const wrapper = mount(BccTable, {
      props: {
        columns,
        items,
      },
      slots: {
        "item.name": "Custom name content",
      },
    });

    expect(wrapper.text()).toContain("Year group");
    expect(wrapper.text()).not.toContain("Ada Lovelace");
    expect(wrapper.text()).toContain("Custom name content");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("can sort a column", async () => {
    const wrapper = mount(BccTable, {
      props: {
        columns,
        items,
      },
    });

    // Click the first column (year group)
    await wrapper.find(".bcc-table-header-sort").trigger("click");

    expect(wrapper.emitted("update:sortBy")?.length).toBe(1);
    expect(wrapper.emitted("update:sortBy")![0]).toEqual(["year_group"]);
    expect(wrapper.emitted("update:sortDirection")![0]).toEqual(["ascending"]);
  });
});
