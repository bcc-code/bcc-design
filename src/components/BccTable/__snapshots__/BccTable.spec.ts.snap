// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BccTable > can render a custom table cell 1`] = `
"<div class=\\"bcc-table-wrapper\\">
  <table class=\\"bcc-table\\">
    <thead>
      <tr class=\\"bcc-table-header-row\\">
        <th class=\\"bcc-table-header bcc-table-header-sortable\\">
          <div class=\\"bcc-table-header-content\\">Year group <button aria-label=\\"Sort column\\" class=\\"bcc-table-header-sort\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" aria-hidden=\\"true\\" viewBox=\\"0 96 960 960\\" fill=\\"currentColor\\" class=\\"bcc-table-header-sort-icon bcc-table-header-sort-icon-undetermined\\">
                <path d=\\"M571.935 855.217V579.261q0-14.674 9.695-24.489 9.696-9.815 24.37-9.815 14.674 0 24.37 9.815 9.695 9.815 9.695 24.489v276.956l95.065-93.826q9.196-9.434 22.99-9.315 13.793.12 23.989 10.315 9.956 10.196 9.956 23.87 0 13.674-10.195 23.87l-151 151q-5.479 5.478-11.696 7.837-6.217 2.358-12.174 2.358-6.717 0-13.054-2.358-6.337-2.359-10.816-7.837l-152-153q-9.195-9.435-9.195-22.99 0-13.554 10.195-23.75 9.196-10.195 23.37-9.695 14.174.5 24.37 10.695l92.065 91.826Zm-254-560.673-93.065 93.065q-9.196 9.195-22.87 9.195-13.674 0-23.87-10.195-10.195-9.957-10.195-23.75 0-13.794 10.195-23.99l151-151q5.479-5.478 11.316-7.837 5.837-2.358 12.554-2.358 6.717 0 12.554 2.358 5.837 2.359 11.316 7.837l152 152q9.195 9.196 9.195 22.99 0 13.793-10.195 23.75-10.196 10.195-23.87 10.195-13.674 0-23.87-10.195l-94.065-93.065v277.195q0 14.674-9.695 24.37-9.696 9.695-24.37 9.695-14.674 0-24.37-9.695-9.695-9.696-9.695-24.37V294.544Z\\"></path>
              </svg>
              <!--v-if-->
              <!--v-if-->
            </button></div>
        </th>
        <th class=\\"bcc-table-header bcc-table-header-sortable\\">
          <div class=\\"bcc-table-header-content\\">Name <button aria-label=\\"Sort column\\" class=\\"bcc-table-header-sort\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" aria-hidden=\\"true\\" viewBox=\\"0 96 960 960\\" fill=\\"currentColor\\" class=\\"bcc-table-header-sort-icon bcc-table-header-sort-icon-undetermined\\">
                <path d=\\"M571.935 855.217V579.261q0-14.674 9.695-24.489 9.696-9.815 24.37-9.815 14.674 0 24.37 9.815 9.695 9.815 9.695 24.489v276.956l95.065-93.826q9.196-9.434 22.99-9.315 13.793.12 23.989 10.315 9.956 10.196 9.956 23.87 0 13.674-10.195 23.87l-151 151q-5.479 5.478-11.696 7.837-6.217 2.358-12.174 2.358-6.717 0-13.054-2.358-6.337-2.359-10.816-7.837l-152-153q-9.195-9.435-9.195-22.99 0-13.554 10.195-23.75 9.196-10.195 23.37-9.695 14.174.5 24.37 10.695l92.065 91.826Zm-254-560.673-93.065 93.065q-9.196 9.195-22.87 9.195-13.674 0-23.87-10.195-10.195-9.957-10.195-23.75 0-13.794 10.195-23.99l151-151q5.479-5.478 11.316-7.837 5.837-2.358 12.554-2.358 6.717 0 12.554 2.358 5.837 2.359 11.316 7.837l152 152q9.195 9.196 9.195 22.99 0 13.793-10.195 23.75-10.196 10.195-23.87 10.195-13.674 0-23.87-10.195l-94.065-93.065v277.195q0 14.674-9.695 24.37-9.696 9.695-24.37 9.695-14.674 0-24.37-9.695-9.695-9.696-9.695-24.37V294.544Z\\"></path>
              </svg>
              <!--v-if-->
              <!--v-if-->
            </button></div>
        </th>
        <th class=\\"bcc-table-header bcc-table-header-sortable\\">
          <div class=\\"bcc-table-header-content\\">Progress <button aria-label=\\"Sort column\\" class=\\"bcc-table-header-sort\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" aria-hidden=\\"true\\" viewBox=\\"0 96 960 960\\" fill=\\"currentColor\\" class=\\"bcc-table-header-sort-icon bcc-table-header-sort-icon-undetermined\\">
                <path d=\\"M571.935 855.217V579.261q0-14.674 9.695-24.489 9.696-9.815 24.37-9.815 14.674 0 24.37 9.815 9.695 9.815 9.695 24.489v276.956l95.065-93.826q9.196-9.434 22.99-9.315 13.793.12 23.989 10.315 9.956 10.196 9.956 23.87 0 13.674-10.195 23.87l-151 151q-5.479 5.478-11.696 7.837-6.217 2.358-12.174 2.358-6.717 0-13.054-2.358-6.337-2.359-10.816-7.837l-152-153q-9.195-9.435-9.195-22.99 0-13.554 10.195-23.75 9.196-10.195 23.37-9.695 14.174.5 24.37 10.695l92.065 91.826Zm-254-560.673-93.065 93.065q-9.196 9.195-22.87 9.195-13.674 0-23.87-10.195-10.195-9.957-10.195-23.75 0-13.794 10.195-23.99l151-151q5.479-5.478 11.316-7.837 5.837-2.358 12.554-2.358 6.717 0 12.554 2.358 5.837 2.359 11.316 7.837l152 152q9.195 9.196 9.195 22.99 0 13.793-10.195 23.75-10.196 10.195-23.87 10.195-13.674 0-23.87-10.195l-94.065-93.065v277.195q0 14.674-9.695 24.37-9.696 9.695-24.37 9.695-14.674 0-24.37-9.695-9.695-9.696-9.695-24.37V294.544Z\\"></path>
              </svg>
              <!--v-if-->
              <!--v-if-->
            </button></div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"bcc-table-cell\\"><span>22/23</span></td>
        <td class=\\"bcc-table-cell\\">Custom name content</td>
        <td class=\\"bcc-table-cell\\"><span>25</span></td>
        <td></td>
      </tr>
      <tr>
        <td class=\\"bcc-table-cell\\"><span>23/24</span></td>
        <td class=\\"bcc-table-cell\\">Custom name content</td>
        <td class=\\"bcc-table-cell\\"><span>42</span></td>
        <td></td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`BccTable > renders data in a table 1`] = `
"<div class=\\"bcc-table-wrapper\\">
  <table class=\\"bcc-table\\">
    <thead>
      <tr class=\\"bcc-table-header-row\\">
        <th class=\\"bcc-table-header bcc-table-header-sortable\\">
          <div class=\\"bcc-table-header-content\\">Year group <button aria-label=\\"Sort column\\" class=\\"bcc-table-header-sort\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" aria-hidden=\\"true\\" viewBox=\\"0 96 960 960\\" fill=\\"currentColor\\" class=\\"bcc-table-header-sort-icon bcc-table-header-sort-icon-undetermined\\">
                <path d=\\"M571.935 855.217V579.261q0-14.674 9.695-24.489 9.696-9.815 24.37-9.815 14.674 0 24.37 9.815 9.695 9.815 9.695 24.489v276.956l95.065-93.826q9.196-9.434 22.99-9.315 13.793.12 23.989 10.315 9.956 10.196 9.956 23.87 0 13.674-10.195 23.87l-151 151q-5.479 5.478-11.696 7.837-6.217 2.358-12.174 2.358-6.717 0-13.054-2.358-6.337-2.359-10.816-7.837l-152-153q-9.195-9.435-9.195-22.99 0-13.554 10.195-23.75 9.196-10.195 23.37-9.695 14.174.5 24.37 10.695l92.065 91.826Zm-254-560.673-93.065 93.065q-9.196 9.195-22.87 9.195-13.674 0-23.87-10.195-10.195-9.957-10.195-23.75 0-13.794 10.195-23.99l151-151q5.479-5.478 11.316-7.837 5.837-2.358 12.554-2.358 6.717 0 12.554 2.358 5.837 2.359 11.316 7.837l152 152q9.195 9.196 9.195 22.99 0 13.793-10.195 23.75-10.196 10.195-23.87 10.195-13.674 0-23.87-10.195l-94.065-93.065v277.195q0 14.674-9.695 24.37-9.696 9.695-24.37 9.695-14.674 0-24.37-9.695-9.695-9.696-9.695-24.37V294.544Z\\"></path>
              </svg>
              <!--v-if-->
              <!--v-if-->
            </button></div>
        </th>
        <th class=\\"bcc-table-header bcc-table-header-sortable\\">
          <div class=\\"bcc-table-header-content\\">Name <button aria-label=\\"Sort column\\" class=\\"bcc-table-header-sort\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" aria-hidden=\\"true\\" viewBox=\\"0 96 960 960\\" fill=\\"currentColor\\" class=\\"bcc-table-header-sort-icon bcc-table-header-sort-icon-undetermined\\">
                <path d=\\"M571.935 855.217V579.261q0-14.674 9.695-24.489 9.696-9.815 24.37-9.815 14.674 0 24.37 9.815 9.695 9.815 9.695 24.489v276.956l95.065-93.826q9.196-9.434 22.99-9.315 13.793.12 23.989 10.315 9.956 10.196 9.956 23.87 0 13.674-10.195 23.87l-151 151q-5.479 5.478-11.696 7.837-6.217 2.358-12.174 2.358-6.717 0-13.054-2.358-6.337-2.359-10.816-7.837l-152-153q-9.195-9.435-9.195-22.99 0-13.554 10.195-23.75 9.196-10.195 23.37-9.695 14.174.5 24.37 10.695l92.065 91.826Zm-254-560.673-93.065 93.065q-9.196 9.195-22.87 9.195-13.674 0-23.87-10.195-10.195-9.957-10.195-23.75 0-13.794 10.195-23.99l151-151q5.479-5.478 11.316-7.837 5.837-2.358 12.554-2.358 6.717 0 12.554 2.358 5.837 2.359 11.316 7.837l152 152q9.195 9.196 9.195 22.99 0 13.793-10.195 23.75-10.196 10.195-23.87 10.195-13.674 0-23.87-10.195l-94.065-93.065v277.195q0 14.674-9.695 24.37-9.696 9.695-24.37 9.695-14.674 0-24.37-9.695-9.695-9.696-9.695-24.37V294.544Z\\"></path>
              </svg>
              <!--v-if-->
              <!--v-if-->
            </button></div>
        </th>
        <th class=\\"bcc-table-header bcc-table-header-sortable\\">
          <div class=\\"bcc-table-header-content\\">Progress <button aria-label=\\"Sort column\\" class=\\"bcc-table-header-sort\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" aria-hidden=\\"true\\" viewBox=\\"0 96 960 960\\" fill=\\"currentColor\\" class=\\"bcc-table-header-sort-icon bcc-table-header-sort-icon-undetermined\\">
                <path d=\\"M571.935 855.217V579.261q0-14.674 9.695-24.489 9.696-9.815 24.37-9.815 14.674 0 24.37 9.815 9.695 9.815 9.695 24.489v276.956l95.065-93.826q9.196-9.434 22.99-9.315 13.793.12 23.989 10.315 9.956 10.196 9.956 23.87 0 13.674-10.195 23.87l-151 151q-5.479 5.478-11.696 7.837-6.217 2.358-12.174 2.358-6.717 0-13.054-2.358-6.337-2.359-10.816-7.837l-152-153q-9.195-9.435-9.195-22.99 0-13.554 10.195-23.75 9.196-10.195 23.37-9.695 14.174.5 24.37 10.695l92.065 91.826Zm-254-560.673-93.065 93.065q-9.196 9.195-22.87 9.195-13.674 0-23.87-10.195-10.195-9.957-10.195-23.75 0-13.794 10.195-23.99l151-151q5.479-5.478 11.316-7.837 5.837-2.358 12.554-2.358 6.717 0 12.554 2.358 5.837 2.359 11.316 7.837l152 152q9.195 9.196 9.195 22.99 0 13.793-10.195 23.75-10.196 10.195-23.87 10.195-13.674 0-23.87-10.195l-94.065-93.065v277.195q0 14.674-9.695 24.37-9.696 9.695-24.37 9.695-14.674 0-24.37-9.695-9.695-9.696-9.695-24.37V294.544Z\\"></path>
              </svg>
              <!--v-if-->
              <!--v-if-->
            </button></div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"bcc-table-cell\\"><span>22/23</span></td>
        <td class=\\"bcc-table-cell\\"><span>Ada Lovelace</span></td>
        <td class=\\"bcc-table-cell\\"><span>25</span></td>
        <td></td>
      </tr>
      <tr>
        <td class=\\"bcc-table-cell\\"><span>23/24</span></td>
        <td class=\\"bcc-table-cell\\"><span>Firmus Piett</span></td>
        <td class=\\"bcc-table-cell\\"><span>42</span></td>
        <td></td>
      </tr>
    </tbody>
  </table>
</div>"
`;
